File: Samples/ARTCCM-TNIR-sample_EN_ver.1.3.1.3c/Sample/Win32/Delphi2006/Sample_Delphi_2006_FullControl/Unit5.pas
unit Unit5;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, ComCtrls, ArtCam;

type
  TForm5 = class(TForm)
    ComboBox1: TComboBox;
    ComboBox2: TComboBox;
    CheckBox1: TCheckBox;
    CheckBox2: TCheckBox;
    CheckBox3: TCheckBox;
    CheckBox4: TCheckBox;
    Button1: TButton;
    Button2: TButton;
    Button3: TButton;
    TrackBar1: TTrackBar;
    TrackBar2: TTrackBar;
    TrackBar3: TTrackBar;
    TrackBar4: TTrackBar;
    TrackBar5: TTrackBar;
    Edit1: TEdit;
    Edit2: TEdit;
    Edit3: TEdit;
    Edit4: TEdit;
    Edit5: TEdit;
    Button5: TButton;
    GroupBox1: TGroupBox;
    GroupBox2: TGroupBox;
    Label1: TLabel;
    Label2: TLabel;
    Label3: TLabel;
    Label4: TLabel;
    Label5: TLabel;
    procedure Button3Click(Sender: TObject);
    procedure Button5Click(Sender: TObject);
    procedure TrackBar5Change(Sender: TObject);
    procedure TrackBar4Change(Sender: TObject);
    procedure TrackBar3Change(Sender: TObject);
    procedure TrackBar2Change(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure KeyValueCheck(Sender: TObject; var Key: Char);
    procedure Edit1Change(Sender: TObject);
    procedure Edit2Change(Sender: TObject);
    procedure Edit3Change(Sender: TObject);
    procedure Edit4Change(Sender: TObject);
    procedure Edit5Change(Sender: TObject);
    procedure TrackBar1Change(Sender: TObject);
    procedure ComboBox2Change(Sender: TObject);
    procedure ComboBox1Change(Sender: TObject);
    procedure CheckBox4Click(Sender: TObject);
    procedure CheckBox3Click(Sender: TObject);
    procedure CheckBox2Click(Sender: TObject);
    procedure CheckBox1Click(Sender: TObject);
    procedure Button2Click(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure FormShow(Sender: TObject);

  private
    { Private declarations }
  public
    m_pCArtCamSdk: ^TArtCam;

  constructor NilSet;
  
  end;

var
  Form5: TForm5;
  lRed :          LONG;
  lBlue:          LONG;
  lGreen:         LONG;
  lGlobalGain:    LONG;
  lShutterWidth:  LONG;


implementation

{$R *.dfm}

constructor TForm5.NilSet;
begin
  m_pCArtCamSdk := nil;
end;

procedure TForm5.FormCreate(Sender: TObject);
begin
  SetWindowLong(Edit1.Handle,GWL_STYLE,
                GetWindowLong(Edit1.Handle,GWL_STYLE)
                or ES_RIGHT
                or ES_NUMBER);
  SetWindowPos(Edit1.Handle,0,0,0,0,0,
               SWP_NOMOVE or SWP_NOSIZE or
               SWP_NOZORDER or SWP_FRAMECHANGED);

  SetWindowLong(Edit2.Handle,GWL_STYLE,
                GetWindowLong(Edit2.Handle,GWL_STYLE)
                or ES_RIGHT
                or ES_NUMBER);
  SetWindowPos(Edit2.Handle,0,0,0,0,0,
               SWP_NOMOVE or SWP_NOSIZE or
               SWP_NOZORDER or SWP_FRAMECHANGED);

  SetWindowLong(Edit3.Handle,GWL_STYLE,
                GetWindowLong(Edit3.Handle,GWL_STYLE)
                or ES_RIGHT
                or ES_NUMBER);
  SetWindowPos(Edit3.Handle,0,0,0,0,0,
               SWP_NOMOVE or SWP_NOSIZE or
               SWP_NOZORDER or SWP_FRAMECHANGED);

  SetWindowLong(Edit4.Handle,GWL_STYLE,
                GetWindowLong(Edit4.Handle,GWL_STYLE)
                or ES_RIGHT
                or ES_NUMBER);
  SetWindowPos(Edit4.Handle,0,0,0,0,0,
               SWP_NOMOVE or SWP_NOSIZE or
               SWP_NOZORDER or SWP_FRAMECHANGED);

  SetWindowLong(Edit5.Handle,GWL_STYLE,
                GetWindowLong(Edit5.Handle,GWL_STYLE)
                or ES_RIGHT
                or ES_NUMBER);
  SetWindowPos(Edit5.Handle,0,0,0,0,0,
               SWP_NOMOVE or SWP_NOSIZE or
               SWP_NOZORDER or SWP_FRAMECHANGED);
end;

procedure TForm5.TrackBar1Change(Sender: TObject);
begin
    lRed := TrackBar1.Position;
    m_pCArtCamSdk.Monitor_SetBayerGainRed(lRed);
    Edit1.Text := IntToStr(lRed);
end;

procedure TForm5.TrackBar2Change(Sender: TObject);
begin
    lGreen := TrackBar2.Position;
    m_pCArtCamSdk.Monitor_SetBayerGainGreen(lGreen);
    Edit2.Text := IntToStr(lGreen);
end;

procedure TForm5.TrackBar3Change(Sender: TObject);
begin
    lBlue := TrackBar3.Position;
    m_pCArtCamSdk.Monitor_SetBayerGainBlue(lBlue);
    Edit3.Text := IntToStr(lBlue);
end;

procedure TForm5.TrackBar4Change(Sender: TObject);
begin
    lGlobalGain := TrackBar4.Position;
    m_pCArtCamSdk.SetGlobalGain(lGlobalGain);
    Edit4.Text := IntToStr(lGlobalGain);
end;

procedure TForm5.TrackBar5Change(Sender: TObject);
begin
    lShutterWidth := TrackBar5.Position;
    m_pCArtCamSdk.SetExposureTime(lShutterWidth);
    Edit5.Text := IntToStr(lShutterWidth);
end;

procedure TForm5.Button1Click(Sender: TObject);
begin
    m_pCArtCamSdk.Monitor_SetPreviewMode(true);
end;

procedure TForm5.Button2Click(Sender: TObject);
begin
  m_pCArtCamSdk.Monitor_SetPreviewMode(false);
end;

procedure TForm5.Button3Click(Sender: TObject);
var dwVFlip: DWORD;
    wReg0xE4: DWORD;
begin
	//
	lShutterWidth	:= m_pCArtCamSdk.GetExposureTime(nil);
	lGlobalGain		:= m_pCArtCamSdk.GetGlobalGain(nil);
	dwVFlip	:= m_pCArtCamSdk.ReadRegister($20,nil);

	// 
	m_pCArtCamSdk.Fpga_WriteRegister($B6, $03);

	// Shutter
	m_pCArtCamSdk.Fpga_WriteRegister($C0, $09);
	m_pCArtCamSdk.Fpga_WriteRegister($C1, BYTE(lShutterWidth shr 8));
	m_pCArtCamSdk.Fpga_WriteRegister($C2, BYTE(lShutterWidth));

	// Gain
	m_pCArtCamSdk.Fpga_WriteRegister($C3, $35);
	m_pCArtCamSdk.Fpga_WriteRegister($C4, BYTE(lGlobalGain shr 8));
	m_pCArtCamSdk.Fpga_WriteRegister($C5, BYTE(lGlobalGain));

	// Vertical mirror
	m_pCArtCamSdk.Fpga_WriteRegister($C6, $20);
	m_pCArtCamSdk.Fpga_WriteRegister($C7, BYTE(dwVFlip shr 8));
	m_pCArtCamSdk.Fpga_WriteRegister($C8, BYTE(dwVFlip));


	//
  wReg0xE4 := m_pCArtCamSdk.Fpga_ReadRegister($E4,nil);

	wReg0xE4 := wReg0xE4 or $0001;
	m_pCArtCamSdk.Fpga_WriteRegister($E4, wReg0xE4);
	wReg0xE4 := wReg0xE4 and $FFFE;
	m_pCArtCamSdk.Fpga_WriteRegister($E4, wReg0xE4);
                                              

end;

procedure TForm5.Button5Click(Sender: TObject);
begin
  Hide;
end;

procedure TForm5.CheckBox1Click(Sender: TObject);
begin
  m_pCArtCamSdk.Monitor_SetColorMode(CheckBox1.Checked);
end;

procedure TForm5.CheckBox2Click(Sender: TObject);
begin
  if (CheckBox2.Checked) then begin
  m_pCArtCamSdk.Monitor_SetBayerGainAuto(1);
  end
  else begin
  m_pCArtCamSdk.Monitor_SetBayerGainAuto(0);
  end;
end;

procedure TForm5.CheckBox3Click(Sender: TObject);
begin
  if (CheckBox3.Checked)  then begin
    m_pCArtCamSdk.Monitor_SetBayerGainLock(1);
  end
  else begin
    m_pCArtCamSdk.Monitor_SetBayerGainLock(0);
  end;
end;

procedure TForm5.CheckBox4Click(Sender: TObject);
begin
    m_pCArtCamSdk.SetMirrorV(CheckBox4.Checked)
end;

procedure TForm5.ComboBox1Change(Sender: TObject);
begin
  m_pCArtCamSdk.Monitor_SetCameraClock(ComboBox1.ItemIndex);
end;

procedure TForm5.ComboBox2Change(Sender: TObject);
begin
  case ComboBox2.ItemIndex of
    0: begin
       m_pCArtCamSdk.Monitor_SetPreviewSize(800,600);
       end;
    1: begin
      m_pCArtCamSdk.Monitor_SetPreviewSize(1024,768);
    end;
    else begin
      m_pCArtCamSdk.Monitor_SetPreviewSize(1280,1024);
    end;

  end;
end;

procedure TForm5.KeyValueCheck(Sender: TObject; var Key: Char);
begin
  if  (Key < '0') or (Key > '9') then Key:= #0;
end;

procedure TForm5.Edit1Change(Sender: TObject);
begin
  if not (Edit1.Text = '') then begin
    m_pCArtCamSdk.Monitor_SetBayerGainRed(StrToInt(Edit1.Text));
  end;
end;

procedure TForm5.Edit2Change(Sender: TObject);
begin
  if not (Edit2.Text = '') then begin
    m_pCArtCamSdk.Monitor_SetBayerGainGreen(StrToInt(Edit2.Text));
  end;
end;

procedure TForm5.Edit3Change(Sender: TObject);
begin
  if not (Edit3.Text = '') then begin
    m_pCArtCamSdk.Monitor_SetBayerGainBlue(StrToInt(Edit3.Text));
  end;
end;

procedure TForm5.Edit4Change(Sender: TObject);
begin
  if not (Edit4.Text = '') then begin
    m_pCArtCamSdk.SetGlobalGain(StrToInt(Edit4.Text));
  end;
end;

procedure TForm5.Edit5Change(Sender: TObject);
begin
  if not (Edit5.Text = '') then begin
    m_pCArtCamSdk.SetExposureTime(StrToInt(Edit5.Text));
  end;
end;

procedure TForm5.FormShow(Sender: TObject);
var
  pWidth:  integer;
  pHeight: integer;
begin
  if nil = m_pCArtCamSdk then begin
    exit;
  end;

  case m_pCArtCamSdk.Monitor_GetCameraClock(nil) of
     2: begin
        ComboBox1.ItemIndex := 1;
     end;
     5: begin
        ComboBox1.ItemIndex := 2;
     end;

     else begin
        ComboBox1.ItemIndex := 0;
     end;
  end;

  m_pCArtCamSdk.Monitor_GetPreviewSize(@pWidth,@pHeight);
    case pWidth of
      800: begin
        ComboBox2.ItemIndex :=0;
      end;
      1024: begin
        ComboBox2.ItemIndex :=1;
      end;
      else begin
        ComboBox2.ItemIndex :=2;
      end;
    end;

  CheckBox1.Checked := m_pCArtCamSdk.Monitor_GetColorMode(nil);

  if (m_pCArtCamSdk.Monitor_GetBayerGainAuto(nil) = 1) then begin
    CheckBox2.Checked :=true;
  end
  else begin
    CheckBox2.Checked := false;
  end;

  if (m_pCArtCamSdk.Monitor_GetBayerGainLock(nil) = 1) then begin
    CheckBox3.Checked :=true;
  end
  else begin
    CheckBox3.Checked := false;
  end;

   CheckBox4.Checked := m_pCArtCamSdk.GetMirrorV();


   lRed  := m_pCArtCamSdk.Monitor_GetBayerGainRed(nil);
   lBlue := m_pCArtCamSdk.Monitor_GetBayerGainBlue(nil);
   lGreen:= m_pCArtCamSdk.Monitor_GetBayerGainGreen(nil);

   TrackBar1.Position := lRed;
   TrackBar2.Position := lGreen;
   TrackBar3.Position := lBlue;

   Edit1.Text := IntToStr(lRed);
   Edit2.Text := IntToStr(lGreen);
   Edit3.Text := IntToStr(lBlue);


   lGlobalGain   := m_pCArtCamSdk.GetGlobalGain(nil);
   lShutterWidth := m_pCArtCamSdk.GetExposureTime(nil);

   TrackBar4.Position := lGlobalGain;
   TrackBar5.Position := lShutterWidth;

   Edit4.Text := IntToStr(lGlobalGain);
   Edit5.Text := IntToStr(lShutterWidth);



  end;

end.

File: Samples/ARTCCM-TNIR-sample_EN_ver.1.3.1.3c/Sample/Win32/BCB2006/Sample_BCB_2006_FullControl/Unit5.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Unit1.h"
#include "Unit5.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TForm5 *Form5;
//---------------------------------------------------------------------------
__fastcall TForm5::TForm5(TComponent* Owner)
	: TForm(Owner)
{
	m_pCArtCamSdk = NULL;
}

//---------------------------------------------------------------------------


void __fastcall TForm5::FormShow(TObject *Sender)
{
	if(NULL == m_pCArtCamSdk){
		return;
	}

	if(m_pCArtCamSdk->Monitor_GetColorMode()){
		CheckBox1->Checked = TRUE;
	}else{
		CheckBox2->Enabled = FALSE;
		CheckBox3->Enabled = FALSE;
	}

	if(m_pCArtCamSdk->Monitor_GetBayerGainAuto()){
		CheckBox2->Checked = TRUE;


		TrackBar1->Enabled = FALSE;
		TrackBar2->Enabled = FALSE;
		TrackBar3->Enabled = FALSE;

		Edit1->Enabled = FALSE;
		Edit2->Enabled = FALSE;
		Edit3->Enabled = FALSE;

	}else{
		CheckBox3->Enabled = FALSE;
	}

	if(m_pCArtCamSdk->Monitor_GetBayerGainLock()){
		CheckBox3->Checked = TRUE;
	}

	if(m_pCArtCamSdk->GetMirrorV()){
    	CheckBox4->Checked = TRUE;
	}

	LONG lWidth = 0; LONG lHeight = 0;
	m_pCArtCamSdk->Monitor_GetPreviewSize(&lWidth,&lHeight);
	switch(lWidth){
		case  800: ComboBox2->ItemIndex=0; break;
		case 1024: ComboBox2->ItemIndex=1; break;
		case 1280: ComboBox2->ItemIndex=2; break;
	}

	LONG lCameraClock = m_pCArtCamSdk->Monitor_GetCameraClock();
		switch(lCameraClock){
			case 2:   ComboBox1->ItemIndex=1; break;
			case 5:   ComboBox1->ItemIndex=2; break;
			default:  ComboBox1->ItemIndex=0; break;
		}

	// Bayer
	lRed   = m_pCArtCamSdk->Monitor_GetBayerGainRed();
	lGreen = m_pCArtCamSdk->Monitor_GetBayerGainGreen();
	lBlue  = m_pCArtCamSdk->Monitor_GetBayerGainBlue();

	TrackBar1->Position = lRed;
	TrackBar2->Position = lGreen;
	TrackBar3->Position = lBlue;

	lGlobalGain   = m_pCArtCamSdk->GetGlobalGain();
	lShutterWidth = m_pCArtCamSdk->GetExposureTime();

	TrackBar4->Position = lGlobalGain;
	TrackBar5->Position = lShutterWidth;


	return;
  
}







void __fastcall TForm5::ComboBox1Change(TObject *Sender)
{
	m_pCArtCamSdk->Monitor_SetCameraClock(ComboBox1->ItemIndex);
}
//---------------------------------------------------------------------------

void __fastcall TForm5::ComboBox2Change(TObject *Sender)
{
	LONG lWidth = 0; LONG lHeight = 0;
	switch(ComboBox2->ItemIndex){
		case 0: lWidth = 800; lHeight = 600; break;
		case 1: lWidth = 1024; lHeight = 768; break;
		default: lWidth = 1280; lHeight = 1024; break;
	}
	m_pCArtCamSdk->Monitor_SetPreviewSize(lWidth,lHeight);
}
//---------------------------------------------------------------------------

void __fastcall TForm5::Button1Click(TObject *Sender)
{
	m_pCArtCamSdk->Monitor_SetPreviewMode(TRUE);	
}
//---------------------------------------------------------------------------

void __fastcall TForm5::Button2Click(TObject *Sender)
{
	m_pCArtCamSdk->Monitor_SetPreviewMode(FALSE);	
}
//---------------------------------------------------------------------------


void TForm5::ColorCheck()
{
	BOOL Color = CheckBox1->Checked;
	BOOL AWB = CheckBox2->Checked;
	BOOL Lock = CheckBox3->Checked;

	m_pCArtCamSdk->Monitor_SetColorMode(Color);
	m_pCArtCamSdk->Monitor_SetBayerGainAuto(AWB);
	m_pCArtCamSdk->Monitor_SetBayerGainLock(Lock);


	TrackBar1->Enabled = (Color && !AWB);
	TrackBar2->Enabled = (Color && !AWB);
	TrackBar3->Enabled = (Color && !AWB);

	Edit1->Enabled = (Color && !AWB);
	Edit2->Enabled = (Color && !AWB);
	Edit3->Enabled = (Color && !AWB);

	if(!Color){
		CheckBox2->Enabled = FALSE;
		CheckBox3->Enabled = FALSE;
	}else{
		CheckBox2->Enabled = TRUE;;
		if(!AWB){
			CheckBox3->Enabled = FALSE;
		}else{
			CheckBox3->Enabled = TRUE;
		}
	}
}
void __fastcall TForm5::CheckBox1Click(TObject *Sender)
{
	ColorCheck();
}
//---------------------------------------------------------------------------

void __fastcall TForm5::CheckBox2Click(TObject *Sender)
{
	ColorCheck();	
}
//---------------------------------------------------------------------------

void __fastcall TForm5::CheckBox3Click(TObject *Sender)
{
	ColorCheck();	
}
//---------------------------------------------------------------------------

void __fastcall TForm5::CheckBox4Click(TObject *Sender)
{
	if( CheckBox4->Checked == TRUE){
		m_pCArtCamSdk->SetMirrorV(1);
	}else{
		m_pCArtCamSdk->SetMirrorV(0);
	}
}
//---------------------------------------------------------------------------

void __fastcall TForm5::TrackBar1Change(TObject *Sender)
{
	lRed = TrackBar1->Position;
	m_pCArtCamSdk->Monitor_SetBayerGainRed(lRed);
	Edit1->Text = IntToStr(lRed);
}
//---------------------------------------------------------------------------

void __fastcall TForm5::TrackBar2Change(TObject *Sender)
{
	lGreen = TrackBar2->Position;
	m_pCArtCamSdk->Monitor_SetBayerGainGreen(lGreen);
	Edit2->Text = IntToStr(lGreen);
}
//---------------------------------------------------------------------------

void __fastcall TForm5::TrackBar3Change(TObject *Sender)
{
	lBlue = TrackBar3->Position;
	m_pCArtCamSdk->Monitor_SetBayerGainBlue(lBlue);
	Edit3->Text = IntToStr(lBlue);
}
//---------------------------------------------------------------------------

void __fastcall TForm5::TrackBar4Change(TObject *Sender)
{
	lGlobalGain = TrackBar4->Position;
	m_pCArtCamSdk->SetGlobalGain(lGlobalGain);
	Edit4->Text = IntToStr(lGlobalGain);
}
//---------------------------------------------------------------------------

void __fastcall TForm5::TrackBar5Change(TObject *Sender)
{
	lShutterWidth = TrackBar5->Position;
	m_pCArtCamSdk->SetExposureTime(lShutterWidth);
	Edit5->Text = IntToStr(lShutterWidth);
}
//---------------------------------------------------------------------------

void __fastcall TForm5::Edit1Change(TObject *Sender)
{
	if(Edit1->Text != ""){
		lRed = StrToInt(Edit1->Text);
		m_pCArtCamSdk->Monitor_SetBayerGainRed(lRed);
		TrackBar1->Position = lRed;
	}
}
//---------------------------------------------------------------------------


void __fastcall TForm5::KeyValueCheck(TObject *Sender, char &Key)
{
	if ((Key<'0') || ('9'<Key))   Key = 0;
}
//---------------------------------------------------------------------------

void __fastcall TForm5::Edit2Change(TObject *Sender)
{
	if(Edit2->Text != ""){
		lGreen = StrToInt(Edit2->Text);
		m_pCArtCamSdk->Monitor_SetBayerGainGreen(lGreen);
		TrackBar2->Position = lGreen;
	}
}
//---------------------------------------------------------------------------

void __fastcall TForm5::Edit3Change(TObject *Sender)
{
	if(Edit3->Text != ""){
		lBlue = StrToInt(Edit3->Text);
		m_pCArtCamSdk->Monitor_SetBayerGainBlue(lBlue);
		TrackBar3->Position = lBlue;
	}
}
//---------------------------------------------------------------------------

void __fastcall TForm5::Edit4Change(TObject *Sender)
{
	if(Edit4->Text != ""){
		lGlobalGain = StrToInt(Edit4->Text);
		m_pCArtCamSdk->SetGlobalGain(lGlobalGain);
		TrackBar4->Position = lGlobalGain;
	}
}
//---------------------------------------------------------------------------

void __fastcall TForm5::Edit5Change(TObject *Sender)
{
	if(Edit5->Text != ""){
		lShutterWidth = StrToInt(Edit5->Text);
		m_pCArtCamSdk->SetExposureTime(lShutterWidth);
		TrackBar5->Position = lShutterWidth;
	}
}
//---------------------------------------------------------------------------

void __fastcall TForm5::Button5Click(TObject *Sender)
{
	Hide();	
}
//---------------------------------------------------------------------------

void __fastcall TForm5::Button3Click(TObject *Sender)
{

	lShutterWidth	= m_pCArtCamSdk->GetExposureTime();
	lGlobalGain		= m_pCArtCamSdk->GetGlobalGain();
	DWORD dwVFlip	= m_pCArtCamSdk->ReadRegister(0x20);


	m_pCArtCamSdk->Fpga_WriteRegister(0xB6, 0x03);

	// Shutter
	m_pCArtCamSdk->Fpga_WriteRegister(0xC0, 0x09);
	m_pCArtCamSdk->Fpga_WriteRegister(0xC1, BYTE(lShutterWidth >> 8));
	m_pCArtCamSdk->Fpga_WriteRegister(0xC2, BYTE(lShutterWidth));

	// Gain
	m_pCArtCamSdk->Fpga_WriteRegister(0xC3, 0x35);
	m_pCArtCamSdk->Fpga_WriteRegister(0xC4, BYTE(lGlobalGain >> 8));
	m_pCArtCamSdk->Fpga_WriteRegister(0xC5, BYTE(lGlobalGain));

	// Mirror reversal
	m_pCArtCamSdk->Fpga_WriteRegister(0xC6, 0x20);
	m_pCArtCamSdk->Fpga_WriteRegister(0xC7, BYTE(dwVFlip >> 8));
	m_pCArtCamSdk->Fpga_WriteRegister(0xC8, BYTE(dwVFlip));


	DWORD wReg0xE4 = m_pCArtCamSdk->Fpga_ReadRegister(0xE4);

	wReg0xE4 |= 0x0001;
	m_pCArtCamSdk->Fpga_WriteRegister(0xE4, wReg0xE4);
	wReg0xE4 &= 0xFFFE;
	m_pCArtCamSdk->Fpga_WriteRegister(0xE4, wReg0xE4);
}
//---------------------------------------------------------------------------

